### PR TITLE
types of joins in sql

### PR DESCRIPTION
**Pull request by kokosing**
types of joins in sql

### CODE REVIEW

**Comment by kokosing** [comment]
@nat925, hej, mogę prosić o review :)

File: _posts/2016-12-31-the-fundamentals-types-of-joins-in-sql.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
+---
+title: 'The fundamentals: types of joins in SQL'
+category: internals
+tags: join fundamentals
+---
+
+Joins are one of the most important part of each database and SQL itself. So it was obvious to me that this topic is going to appear very often on this blog.
```
Comment (nat925) [inline _posts/2016-12-31-the-fundamentals-types-of-joins-in-sql.md]: part -> parts

File: _posts/2016-12-31-the-fundamentals-types-of-joins-in-sql.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
+---
+title: 'The fundamentals: types of joins in SQL'
+category: internals
+tags: join fundamentals
+---
+
+Joins are one of the most important part of each database and SQL itself. So it was obvious to me that this topic is going to appear very often on this blog. 
+Hovewer, I thought that my first technical article will be about something more advanced, like join reordering or at least cross join elimination. 
+Although, when thinking about what I could write about I realized that I would need to explain the basic terminology, so we could find a common domain language. 
+Probably, most of below you know already, but even so I hope this is going to help you structurize your knowledge and find the missing things.
```
Comment (nat925) [inline _posts/2016-12-31-the-fundamentals-types-of-joins-in-sql.md]: most of below -> most of the things below

File: _posts/2016-12-31-the-fundamentals-types-of-joins-in-sql.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+It is the first post in the series where I going to explain fundamental knowledge about database internals. So please do not be scarred that is so entry level, hope next posts will be more challenging for you.
```
Comment (nat925) [inline _posts/2016-12-31-the-fundamentals-types-of-joins-in-sql.md]: I going -> I am going
scarred -> scared
hope -> hopefully

File: _posts/2016-12-31-the-fundamentals-types-of-joins-in-sql.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+Lets start from the ANSI SQL. What do we have there in regards of joins? There are following explicit types of join:
```
Comment (nat925) [inline _posts/2016-12-31-the-fundamentals-types-of-joins-in-sql.md]: regards of -> regards to

File: _posts/2016-12-31-the-fundamentals-types-of-joins-in-sql.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+* cross join - the simples one, two tables are joined together without any condition. Every single row from the one table will be joined with every row from the other table - a cartesian prudct [^join_syntax]:
```
Comment (nat925) [inline _posts/2016-12-31-the-fundamentals-types-of-joins-in-sql.md]: simples one -> simple ones
a cartesian prudct -> podejrzewam, że chodziło o product :)

File: _posts/2016-12-31-the-fundamentals-types-of-joins-in-sql.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+presto:tiny> SELECT * FROM (VALUES 1, 2) t("left"), (VALUES 1, 1, 2) u("right") WHERE t."left" = u."right";
+ left | right 
+------+-------
+    1 |     1 
+    1 |     1 
+    2 |     2 
+(3 rows)
+~~~
+
+* left outer join - like inner join, but when pair of rows does not satisfy the join condition, the left row completed with NULL values is returned [^join_syntax]:
```
Comment (nat925) [inline _posts/2016-12-31-the-fundamentals-types-of-joins-in-sql.md]: pair of rows -> a pair of rows

File: _posts/2016-12-31-the-fundamentals-types-of-joins-in-sql.md [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    2 | NULL
+ NULL |     3
+(3 rows)
+~~~
+
+There are also implicit join types. Joins which do not have such a good syntax representation in SQL:
+
+ * semi join - a row from the one table is returned when there is a row in the other table for which a join condition is met. You might think that it is not a join, as only a row from a single table is returned, but it is called join as it is using the other table to evaluate join condition. 
+That is the reason why it is called a semi join. 
+There are several ways to express such join in SQL each of them is not direct, may depend on data you have and so its usage could be non trivial.
```
Comment (nat925) [inline _posts/2016-12-31-the-fundamentals-types-of-joins-in-sql.md]: SQL each -> SQL, however each of them is not direct and may depend on the data you have, so its usage could be non trivial.

File: _posts/2016-12-31-the-fundamentals-types-of-joins-in-sql.md [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+ NULL |     3
+(3 rows)
+~~~
+
+There are also implicit join types. Joins which do not have such a good syntax representation in SQL:
+
+ * semi join - a row from the one table is returned when there is a row in the other table for which a join condition is met. You might think that it is not a join, as only a row from a single table is returned, but it is called join as it is using the other table to evaluate join condition. 
+That is the reason why it is called a semi join. 
+There are several ways to express such join in SQL each of them is not direct, may depend on data you have and so its usage could be non trivial. 
+The easiestway is to use `IN` predicate [^simplicity]:
```
Comment (nat925) [inline _posts/2016-12-31-the-fundamentals-types-of-joins-in-sql.md]: brakuje spacji po easiest

File: _posts/2016-12-31-the-fundamentals-types-of-joins-in-sql.md [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+~~~ sql
+presto:tiny> SELECT * FROM (VALUES 1, 2) t(one) WHERE t.one IN (VALUES 1, 1, 3);
+ one 
+-----
+   1 
+(1 row)
+~~~
+
+ * anti join - a very similar to semi join. A row from the one table is returned when the there is no row in the other table for which a condition could be satisfied. Here, as with semi join, there are also several ways express such join.
```
Comment (nat925) [inline _posts/2016-12-31-the-fundamentals-types-of-joins-in-sql.md]: ways express -> ways to express

File: _posts/2016-12-31-the-fundamentals-types-of-joins-in-sql.md [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+The easiest is to use `NOT IN` [^simplicity].
+
+~~~ sql
+presto:tiny> SELECT * FROM (VALUES 1, 2) t(one) WHERE t.one NOT IN (VALUES 1, 1, 3);
+ one 
+-----
+   2 
+~~~
+
+The thing that the semi and anti joins do not have a proper syntax in SQL is somehow bitter. This causes that only a conjunction of equality comparisons can be used as join condition, while for other user does not have such limitation. You cannot express anything like: `SELECT * FROM t SEMI JOIN u ON t.x > u.y`, instead you need to use a correlated subquery, which is a topic for another post.
```
Comment (nat925) [inline _posts/2016-12-31-the-fundamentals-types-of-joins-in-sql.md]: The thing -> może raczej the fact
for other -> in other cases the user

File: _posts/2016-12-31-the-fundamentals-types-of-joins-in-sql.md [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+~~~ sql
+presto:tiny> SELECT * FROM (VALUES 1, 2, 4) t(one) WHERE EXISTS(SELECT * FROM (VALUES 1, 1, 3) t(other) WHERE one > other);
+ one 
+-----
+   2 
+   4 
+(2 rows)
+~~~
+
+That could be all. Nice and easy post to start the journey into Presto internals. In the next post I am going to cover what are join types in the Presto execution engine. I will describe the ways of how all of the above joins are actually calculated.
```
Comment (nat925) [inline _posts/2016-12-31-the-fundamentals-types-of-joins-in-sql.md]: could -> would 
Nice -> A nice 
the ways of how -> wystarczy tylko "how"

File: _posts/2016-12-31-the-fundamentals-types-of-joins-in-sql.md [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+
+[^join_syntax]:
+    Notice that SQL syntax gives to possible ways of to the cross and the inner joins. One is to simply put tables separated with comas, and the other is tables are separated with `JOIN` keyword.
+    In case of the inner join, join condition can be placed after `ON` (when `JOIN` separator is used) keyword or within a `WHERE` clause.
+    In regards of any `OUTER` join `JOIN` followed by `ON` keywords has to be used. This due the fact that join condition has to be explicitly pointed.
+
+[^simplicity]:
+    The easiest in this context means, that it is the easiest for user, though it may not be the most efficient way for data base to execute that.
```
Comment (nat925) [inline _posts/2016-12-31-the-fundamentals-types-of-joins-in-sql.md]: bez przecinka po means (przed that)

File: _posts/2016-12-31-the-fundamentals-types-of-joins-in-sql.md [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+-----
+   2 
+   4 
+(2 rows)
+~~~
+
+That could be all. Nice and easy post to start the journey into Presto internals. In the next post I am going to cover what are join types in the Presto execution engine. I will describe the ways of how all of the above joins are actually calculated.
+
+[^join_syntax]:
+    Notice that SQL syntax gives to possible ways of to the cross and the inner joins. One is to simply put tables separated with comas, and the other is tables are separated with `JOIN` keyword.
```
Comment (nat925) [inline _posts/2016-12-31-the-fundamentals-types-of-joins-in-sql.md]: to -> two
of to -> for
One is to simply put tables separated with comas, and the other is tables are separated with `JOIN` keyword. -> One is to simply put tables separated with commas, the other - with `JOIN` keyword.

File: _posts/2016-12-31-the-fundamentals-types-of-joins-in-sql.md [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+    In case of the inner join, join condition can be placed after `ON` (when `JOIN` separator is used) keyword or within a `WHERE` clause.
+    In regards of any `OUTER` join `JOIN` followed by `ON` keywords has to be used. This due the fact that join condition has to be explicitly pointed.
```
Comment (nat925) [inline _posts/2016-12-31-the-fundamentals-types-of-joins-in-sql.md]: This due -> This is due

**Pull request merged**